ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -89,9 +89,9 @@
    * startTokenBeforeLogFlush: This token is set to the current start token just before log flush and once the log is
    *                           flushed, this is used to set startTokenSafeToPersist.
    */
-  private FindToken startToken;
   private FindToken startTokenBeforeLogFlush;
-  private FindToken startTokenSafeToPersist;
+  private volatile FindToken startTokenSafeToPersist;
```

Comment:
For concurrency, let's make sure all variables shared by HardDeleter thread and Index persistor thread to be volatile.

Reply:
sure.